src/coprocessor/codec/mysql/json/binary.rs

```diff
@@ -223,126 +222,125 @@ pub trait JsonEncoder: NumberEncoder {
-            TYPE_CODE_LITERAL => self.decode_json_literal(),
-            TYPE_CODE_I64 => self.decode_json_i64(),
-            TYPE_CODE_U64 => self.decode_json_u64(),
-            TYPE_CODE_DOUBLE => self.decode_json_double(),
-            TYPE_CODE_STRING => self.decode_json_str(),
-            _ => Err(invalid_type!("unsupported type {:?}", code_type)),
+            TYPE_CODE_OBJECT => Json::decode_obj(buf),
+            TYPE_CODE_ARRAY => Json::decode_array(buf),
+            TYPE_CODE_LITERAL => Json::decode_literal(buf),
+            TYPE_CODE_I64 => Json::decode_i64(buf),
+            TYPE_CODE_U64 => Json::decode_u64(buf),
+            TYPE_CODE_DOUBLE => Json::decode_double(buf),
+            TYPE_CODE_STRING => Json::decode_str(buf),
+            _ => Err(Error::InvalidDataType("unsupported type".into())),
         }
     }
 
-    fn decode_json_obj(&mut self) -> Result<Json> {
+    fn decode_obj(buf: &mut BytesSlice) -> Result<Json> {
         // count size key_entries value_entries keys values
-        let element_count = self.decode_u32_le()? as usize;
-        let total_size = self.decode_u32_le()? as usize;
+        let element_count = number::decode_u32_le(buf)? as usize;
+        let total_size = number::decode_u32_le(buf)? as usize;
         let left_size = total_size - ELEMENT_COUNT_LEN - SIZE_LEN;
-        let mut data = vec![0; left_size];
-        self.read_exact(&mut data)?;
         let mut obj = BTreeMap::new();
         if element_count == 0 {
             return Ok(Json::Object(obj));
         }
         // key_entries
         let key_entries_len = KEY_ENTRY_LEN * element_count;
-        let mut key_entries_data = &data[0..key_entries_len];
+        let mut key_entries_data = read_slice(buf, key_entries_len)?; //&data[0..key_entries_len];
 
         // value-entry ::= type(byte) offset-or-inlined-value(uint32)
         let value_entries_len = VALUE_ENTRY_LEN * element_count;
-        let mut value_entries_data = &data[key_entries_len..(key_entries_len + value_entries_len)];
-        let mut key_offset = key_entries_len + value_entries_len;
+        let mut value_entries_data = read_slice(buf, value_entries_len)?;
+        let mut data = read_slice(buf, left_size - key_entries_len - value_entries_len)?;
         for _ in 0..element_count {
             let key_real_offset = number::decode_u32_le(&mut key_entries_data)?;
             let key_len = number::decode_u16_le(&mut key_entries_data)?;
-            let key_data = &data[key_offset..(key_offset + key_len as usize)];
+            let key_data = read_slice(&mut data, key_len as usize)?;
             let key = String::from(str::from_utf8(key_data).unwrap());
-            let value = value_entries_data.decode_json_item(&data[key_offset..], key_real_offset)?;
+            let value = Json::decode_item(
+                &mut value_entries_data,
+                data,
+                key_real_offset + u32::from(key_len),
```

Comment:
why do we need `+ u32::from(key_len)`?

Reply:
Since L265 has already moved the data forward `key_len`

Reply:
oh. I got it.